File: data-structure/python/lists/exercise-linked-stack.py
from node import Node

class LinkedStack:
    def __init__ (self):
        self.head = None
        self.__top = None

    def __is_empty(self):
        return self.__top == None
    
    def push(self, value):
        n = Node(value)
        if self.__is_empty():
            self.head = n
            self.__top = n
        else:
            self.__top.next = n
            self.__top = n

    def top(self):
        value = self.__top.value
        print(value)
        return value

    def pop(self):
        if self.__top == None:
            print('The stack is empty')
            return
        
        if self.head == self.__top:
            value = self.__top.value
            self.head = None
            self.__top = None
            return value
        
        current = self.head

        while current.next != self.__top:
            current = current.next
        
        temp = self.__top.value
        self.__top = current
        self.__top.next = None

        return temp
        

ls = LinkedStack()
ls.push(1)
ls.push(2)
ls.push(3)
ls.push(4)
ls.push(5)
ls.top()
ls.pop()
ls.top()
ls.pop()
ls.top()
ls.pop()
ls.top()
ls.pop()
ls.top()
ls.pop()
ls.pop() 

File: data-structure/js/lists/exercise-linked-stack.ts
import { Node } from "./node";

class LinkedStack {
  private head: Node | null = null;
  private tail: Node | null = null;

  private isEmpty() {
    return this.head === null;
  }
  public show() {
    if (this.isEmpty()) {
      console.log("The stack is empty");
      return;
    }

    let current = this.head;

    while (current) {
      console.log(current.value);
      current = current.next;
    }
  }
  public push(value: number) {
    const node = new Node(value);

    if (this.isEmpty()) {
      this.head = node;
      this.tail = node;
    } else {
      this.tail!.next = node;
      this.tail = node;
    }
  }
  public top() {
    if (this.isEmpty()) {
      console.log("The stack is empty");
      return;
    }

    return this.tail!.value;
  }
  public pop() {
    if (this.isEmpty()) {
      console.log("The stack is empty");
      return;
    }

    const removedValue = this.tail!.value;

    if (this.head === this.tail) {
      this.head = null;
      this.tail = null;
    } else {
      let current = this.head;
      while (current?.next !== this.tail) {
        current = current!.next;
      }
      this.tail = current;
      this.tail!.next = null;
    }
    return removedValue;
  }
}

const ls = new LinkedStack();
ls.show();
ls.push(1);
ls.push(2);
ls.push(3);
ls.push(4);
ls.pop();
ls.show();
